[WIP][FEATURE] Slack notification

StephenKing opened the pull request:
Hi,

I'd really like to have your feedback on this. I think this is pretty dumb.

Do you have a more fancy solution in mind?

- regarding handling "events"
- regarding the messages that are sent
- regarding where functionality is located (does this Slack class need to be aware of the version input?)

I'd try to avoid over-engineering, but this feels very boring compared to the rest of the code.

Thx

StephenKing committed:
[FEATURE] Build causes from JenkinsHelper

StephenKing committed:
[FEATURE] Slack notifications (regular start/stop)

StephenKing committed:
[FEATURE] Notify also for version bump

michaellihs commented on src/org/typo3/chefci/v2/Pipeline.groovy:
```diff
@@ -85,10 +86,14 @@ class Pipeline implements Serializable {
     }
 
     void execute() {
+        (new Slack(script)).buildStart()
```
hm - the `new` makes no sense here... why not having `slack` as a property and doing

    slack = new Slack(script)

in the constructor - so here it's only `slack.buildStart()` and `slack.buildFinish()` afterwards? Think that's a cleaner solution :)

StephenKing replied:
makes sense..

michaellihs commented on src/org/typo3/chefci/v2/stages/Publish.groovy:
```diff
@@ -21,6 +21,8 @@ class Publish extends AbstractStage {
      */
     protected publish() {
 
+        (new Slack(script)).notifyVersionBump()
```
you could think about injecting slack as an object, when a stage is added to a pipeline - like this you can get rid of the `new ...` and just call `slack.notifyVersionBump()`

You can put the `slack` property and the `setSlack(...)` inject method into the abstract class, so you don't blow up the implementations...

michaellihs replied:
generally speaking, you could add all "helpers" to the abstract stages - so you have them in all stage implementations...

StephenKing replied:
Something like this makes a lot of sense to me. Call that like a `NotificationService` thing or so?

michaellihs replied:
Hm - depends on what you are trying to achieve. If you want to have "generic" notifications and a sink (or many sinks) that handles such a notification, you could have a event that gets "thrown" and can register an `EventHandler` with a concrete implementation `SlackEventHandler` that processes such events.

But to keep it simple, I'd just go with the `Slack` class, but instead of instantiating it each time you call a method on it, instantiate it once (e.g. when the stage is created and registered in the Pipeline). Then you can have the instantiation once for the Pipeline (it should stay the same for all stages in the Pipeline) and have it injected to the Stage, once it is registered at the Pipeline...

michaellihs commented on src/org/typo3/chefci/helpers/Slack.groovy:
```diff
@@ -0,0 +1,42 @@
+
+    Slack(script) {
+        this.script = script
+        jenkinsHelper = new JenkinsHelper(script)
```
maybe not important here, but a good rule of thumb: never use `new` in a constructor - rather inject it as a constructor parameter (or inject it with a setter if optional). This makes your class a lot better testable... but ok - here it is not that important...

StephenKing replied:
your feedback is important! having couple of helper things certainly makes it hard to oversee..

michaellihs commented on src/org/typo3/chefci/helpers/JenkinsHelper.groovy:
```diff
@@ -92,4 +92,13 @@ class JenkinsHelper implements Serializable {
+     */
+    @NonCPS
+    List<String> getBuildCauses() {
+        script.currentBuild.rawBuild.getCauses().collect{ it.getShortDescription() }
```
nice! +1

michaellihs commented on src/org/typo3/chefci/helpers/Slack.groovy:
```diff
@@ -0,0 +1,42 @@
+
+    def notifyVersionBump() {
+        send(
+                message: "${script.env.JOB_NAME} build #${script.env.BUILD_NUMBER} *waiting for input*:\n${script.env.BUILD_URL}",
```
If I get it right, this sends a notification on Slack, that a job is waiting for an input? With a URL to be clicked so you can directly trigger or cancel?!? So nice!!!!

michaellihs reviewed (changes requested):
see detailed comments - all just suggestions for improvements... basically great work!

michaellihs commented:
You don't have to make it more complicated, you can just do a little clean up... if you think you need something like events, you can add them any time... for now I think this looks good as it is with some minor improvements...

StephenKing commented:
Thanks for your feedback. I appreciate it a lot. The current implementation feels very bad to me, so I definitely like to have an improved version. I can't yet imagine all the things. If you find time to create a draft, I'm happy to see that, otherwise I try to wind my head around that next week.

michaellihs committed:
Inject Slack and JenkinsHelper in Stages

Please review carefully, I had no test environment at hand, so this comes without any warranty ;(

michaellihs committed:
Simplify naming of Stages

As we pass the stage name to the abstract constructor, we don't have to set it in the Pipeline Builder - which makes things a little easier...

michaellihs commented:
Hi @StephenKing - I tried to refactor the code to change the things I mentioned. Please review it carefully since I had no testing environment - so it was blind flight :)

See https://github.com/TYPO3-infrastructure/jenkins-pipeline-global-library-chefci/commit/1bc500cdd196b5f4a35c0ed364826b59d4c915d8

and https://github.com/TYPO3-infrastructure/jenkins-pipeline-global-library-chefci/commit/277e4257ef43f76265254d9237d32303197ff65b

Feel free to ask if there are any questions.

P.S. sorry for messing up the commit messages :( Feel free to change them if you want... I was afraid of breaking things if I do it...

StephenKing commented:
Oh yeah, it's talking to us!

![screen shot 2017-01-28 at 20 13 19](https://cloud.githubusercontent.com/assets/277561/22399110/57458092-e596-11e6-869f-5f52681d222c.png)
